### ISSUE TITLE
map readability: labels intersecting each other

### ISSUE THREAD

**Issue by rinigus** [Framstag/libosmscout#287]
I think we can majorly improve readability and visual attractiveness of the maps by resolving issues with the labels drawn on the top of each other. I have attached two screenshots to illustrate what I mean (See below). 

It looks to me that the way labels are not taken into account when checking whether to place the label in LabelLayouter::Placelabel. As a result, way labels can be easily placed on top of each other (see example with such cases circled) or intersect with other labels (arrows). 

To make it possible, probably label layout calculations should be able to receive label description from a label consisting of several turned characters linked with each other. As soon as one is dropped out, all linked chars should be out as well.

Any help, pointers on how to solve it, are very welcome. 

![double-roads](https://cloud.githubusercontent.com/assets/15706605/24334693/af808e60-1277-11e7-8539-5d1a43f1c6c4.png)
![roads-vs-labels](https://cloud.githubusercontent.com/assets/15706605/24334692/af8034f6-1277-11e7-82a7-7cd56249068d.png)

**Comment by Framstag** [comment]
The current algorithm for point labels only works based on labels having some "safe bounding box" and overlapping of these boxes. If boxes overlap, the box/label with the higher priority wins (if they have same priority both labels are dropped).

Technically the code using a tree-like structure for fast search. It expects boxes to be aligned to the horizontal and vertical axis.

It currently does not handle contour labels since they are not aligned to the axis. If you just expand these boxes to aligned boxes (by making them bigger), the algorithm will likely drop too many labels. Implementing a exact "overlapping test" will be much more expensive on the other hand. Handling more labels will create further performance pressure on the algorithm at the same time.

I thus suggest to not try to "hack" the contour labels into the existing algorithm by find a better data structure than can handle arbitrary orientated boxes. There is some/much literature about this topic, though we do not need a high quality algorithm but a fast and "good enough" one.

**Comment by rinigus** [comment]
I am not sure that there will be too many labels dropped if we expand the current contour labels into the current scheme. All labels need some space around to "breeze" and we do have too many labels to start with. But I guess we would need first some additional support to handle such rejection/acceptance that would involve all labels and symbols (pharmacy, parking, ...). Later it would be possible to tune with other bounding box implementations as well.

As for rejecting the both labels if they intersect - its probably better to keep one (with the smaller ID to make it deterministic). Otherwise we would end up without street names where we have two names drawn on the top of each other.

**Comment by Framstag** [comment]
It should be possible to make some simple measurements by injecting contour labels (with adapted bounding box) into the label layouting algorithm (just for performance measurement). Note though, that the algorithm builds a list of to be rendered labels to be processed after all labels have been injected and processed. The same is true for symbols. So moving contour labels into the label layout algorithm might change the order of contour labels rendered, too (tough this might not be a problem).

Some questions:
* If a label consists of multiple text and a symbol/icon and there is some overlap with some other label...which parts will be dropped? Only the overlapping labels or everything? Would it not make sense to have some fallback to the symbol only?
* I did some test with dropping one label in case of overlap of labels with the same priority (this would also simple to change/revert). I had though the feeling that this generates some instability into the rendering, especially on zooming in/out. On the other hand, they may be labels you will never see, because they are too close to something else.

IMHO we should answer these and possibly other questions first to make sure what we want from the label layouting algorithm and t be able to build the correct data structures.

I plan to work on #103 first, because the instability of shield label rendering on zooming in/out is currently the most annoying label placement problem IMHO and the current algorithm possibly rather expensive, too. But whatever I will do for shield label rendering, the resulting labels will be injected into the general label placing algorithm finally so on improvements for that one it can be worked on in parallel.

**Comment by rinigus** [comment]
Re fallback: I would expect to fallback to symbol only, if its possible / allowed by placing of other labels

Re dropping same priority: maybe if we add IDs as a sub-priority ordering, it would help in this case. For road names that are shown in double that's probably required.

At present, I am mainly in the code reading stage anyway, that's as parallel as it gets. :)

**Comment by Framstag** [comment]
Fallback: See MapPainter::LayoutPointLabels(). As you can see, currently each part of a "label" is either directly draw on screen or put into the label layouter. Order of objects within a label is defined by its "position" (See LabelLayoutDataSorter). This allows some combinations.  It makes sense to move everything to the layouting process - this makes some things simpler. I though about moving the whole "label" (icon plus possibly multiple texts) in one bounding box. So this would be "either all or nothing" for the layouting process and since it reduces the number of objects may make it even quicker. But how to implement the fallback? Move a second variant of the label only containing the icons into the layouter, too? So if the full blown one is too big it takes the second one? How to make sure, that they are not detected as conflicting always (as far as I remember objects with the same labelId are never in conflict)?

**Comment by Framstag** [comment]
Fallback: See MapPainter::LayoutPointLabels(). As you can see, currently each part of a "label" is either directly draw on screen or put into the label layouter. Order of objects within a label is defined by its "position" (See LabelLayoutDataSorter). This allows some combinations.  It makes sense to move everything to the layouting process - this makes some things simpler. I though about moving the whole "label" (icon plus possibly multiple texts) in one bounding box. So this would be "either all or nothing" for the layouting process and since it reduces the number of objects may make it even quicker. But how to implement the fallback? Move a second variant of the label only containing the icons into the layouter, too? So if the full blown one is too big it takes the second one? How to make sure, that they are not detected as conflicting always (as far as I remember objects with the same labelId are never in conflict)?

**Comment by rinigus** [comment]
I would have to read the code to make qualified reply. However, just from thinking about it: 

I was thinking of introducing full and smaller label version that can be flagged and checked as a single unit. This should allow symbol+text and symbol combinations.

We would probably have to make something like linked label that connects contour labels into a unit which is all or nothing (don't want characters to drop out from the full name, do we). 

Keep the guidelines and tips coming...

**Comment by Framstag** [comment]
Would you really inject each letter of each contour label instance as one label into the layouter? The layouter currently builds a tree-like structure to hold information which intersection candidates can be at a given position. That will increase the number of objects in tree by some huger amount. I assume for example a to left to bottom right contour label (one instance of for example "Bahnhofstraße") by a bounding box which borders are aligned to screen axis. The bounding box thus is by some amount bigger than required by overlap test will be much simpler. You can of course do the approximation for each letter...

**Comment by rinigus** [comment]
No, I was thinking to use bounding  box for full label. However, I presume that contour label has some kind of rotation and position data for each character. So, this data has to be kept together with the text to be able to draw it after placement decision is made .

**Comment by Framstag** [comment]
The layouter does not need al these information. Some kind of handle, pointer or index offset would be enough to reference detail information how to render the actual label content. The layouter only does information about the box, the priority and similar stuff...

**Comment by rinigus** [comment]
agreed, that's not needed for layouter. this information is needed at the later stage for drawing . however , we do need to estimate bounding  boxes for labels before layouter. It probably could be relatively crude approximation though.

**Comment by Framstag** [comment]
Yes, though currently the box is exact if passed to the layouter. This is not optimal, because calculating the bounding box of text is expensive, but before I worked if rough estimations and the complexity of the resulting algorithms was neither fast ;-) We should though calculate the bounding box of the same text multiple times. I fixed this for Shield labels (See RegisterPointWayLabel()).

**Comment by Karry** [comment]
> No, I was thinking to use bounding box for full label.

@rinigus , I suggest to use list of bounding boxes - separate for each glyph, with countour labels... 

Consider this example:

![label-layout](https://cloud.githubusercontent.com/assets/309458/24367441/ed830c2e-131c-11e7-8a73-b7ef1c96197a.png)

Bounding boxes of street manes intersects each other and with address labels too... When you take bounding box for each glyph, there is just one (probably) collision...

**Comment by rinigus** [comment]
@Karry, its a good example. I wouldn't mind to get rid of one street name (one would remain), but it would be pity to loose a house number(s) here. Boxes with directions (polygon?) would probably resolve it nicely

**Comment by aquiles2k** [comment]
Just thinking here, but wouldnt also help to give the layouter some margins in which he can adjust the label position in a certain direction and see if it still collides? For the housenumbers e.g. this could be simply moving them some pixel left or right of the oringinal position?

Or am i thinking to simple here?

**Comment by Framstag** [comment]
That is certainly possible. The Layouters interface could get improved in this respect but in principle input coordinates do not need to be equal to the output coordinate. With some simple changes it is also possible to implement different layouting algorithms.
Intelligent label layouting is a common problem and there are a number of blog posts, academic papers and similar for this topic.
The main question is: How fast can it be? Label placement in general may not have hard time constraints but for our use case it certainly has. This is (one of the reasons), why did not yet work and more sophisticated layouting but more on performance optimizations. 
Currently the code is rather slow because the number of labels put into the layouter seems to be much higher than the output numbers (so many labels are dropped). If we could reduce the size of the input in a fast way, we can spend more time optimizing the layout of the rest. See #103 for example (which I moved to "later" because this is an issue where I need a longer time frame in bigger slices to get a solution).

**Comment by Karry** [comment]
Hi. I will try to rewrite label layouter with this problem in mind... Give me few weeks.

**Comment by Karry** [comment]
I am closing this issue. New label layouter is using bitmap canvas to mark pixels occupied by some label. For contour labels, separate glyphs are taken into account. All labels are sorted by priority and check for free space in this bitmap before rendering...

**Comment by rinigus** [comment]
@Karry - congratulations!
